Add flume support

ChrisMandich opened the pull request:
## Description:
Add's sensor for Flume water sensor found here: https://flumetech.com/

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#10624

## Example entry for `configuration.yaml` (if applicable):
```yaml
# Example configuration.yaml entry
sensor:
  # Flume
  - platform: flume
    username: YOUR_FLUME_USERNAME
    password: YOUR_FLUME_PASSWORD
    client_id: YOUR_FLUME_CLIENT_ID
    client_secret: YOUR_FLUME_CLIENT_SECRET
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)
Found here: https://github.com/home-assistant/home-assistant.io/pull/10624

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

ChrisMandich committed:
Add Flume Sensor

Add support for Flume API and sensor

ChrisMandich committed:
Add support for choosing timezone

Timezone is now a required option in configuration.yaml

ChrisMandich committed:
Add Flume to coveragerc and CODEOWNERS

Add flume to be ommited from testing. Add @ChrisMandich as Flume Code owner

ChrisMandich committed:
Revert "Add Flume to coveragerc and CODEOWNERS"

This reverts commit 0b27437a3b13f024332c6498b3bb0d554482e49b.

ChrisMandich committed:
Update manifest.json

Added Documentation URL for Flume to HASS.IO

ChrisMandich committed:
Update manifest.json

Added missing Newline at EOF.

ChrisMandich committed:
Update sensor.py

Update sensor to not required Device ID for configuration. Now loops through all available Type 2 devices and adds them as an entity.

ChrisMandich committed:
Update Manifest, CODEOWNERS, and manifest.json

Resolved errors related to code owners and requirements. Using hassfest and gen_requirements_all

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,288 @@
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_CLIENT_ID): cv.string,
+        vol.Required(CONF_CLIENT_SECRET): cv.string,
+        vol.Required(CONF_TIME_ZONE): cv.string,
```
I think we should get the time zone from Home Assistant.

ChrisMandich replied:
Thank you for the feedback! I have updated the sensor accordingly.

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,288 @@
+    time_zone = config.get(CONF_TIME_ZONE)
+    name = config.get(CONF_NAME)
+
+    FlumeDeviceList = FlumeDevice(username, password, client_id, client_secret)
```
You should use snake_case for your variables, so `flume_device_list` (or `flume_devices`)

ChrisMandich replied:
Thank you for the feedback! I have updated the sensor accordingly.

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,288 @@
+    try:
+        for device in FlumeDeviceList._devices:
+            if device["type"] == 2:
+                Flume = FlumeData(
```
`flume` or `flume_data`

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,288 @@
+        self._bearer = self.getBearer()
+        self._devices = self.getDevices()
+
+    def getToken(self):
```
This function is a duplicate of `FlumeData.getToken()`

ChrisMandich replied:
Thank you for the feedback! I have updated the sensor accordingly.

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,288 @@
+                )
+            )
+
+    def getUserId(self):
```
This function is a duplicate of `FlumeData.getUserId()`

ChrisMandich replied:
Thank you for the feedback! I have updated the sensor accordingly.

ChrisMandich committed:
Update sensor.py

Implemented recommendations from @Quentame. Including time zone from Home Assistant, Updated variable names, and Consolidated duplicate functions.

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,246 @@
+                    SCAN_INTERVAL,
+                )
+                sensor_name = name + " " + device["id"]
+                add_entities([FlumeSensor(flume, sensor_name)], True)
```
Use f-strings instead of string concatenation for `sensor_name` (like `self._owner_fullname = f"{user_info['firstName']} {user_info['lastName']}"`).

You can also put it directly in `add_entities`.

ChrisMandich replied:
This has been completed

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,246 @@
+
+    def __init__(self, Flume, name):
+        """Initialize the Flume sensor."""
+        self.Flume = Flume
```
snake_case for variables

`self.flume = flume`

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,246 @@
+
+    try:
+        for device in flume_devices._devices:
+            if device["type"] == 2:
```
Add `2` as the constants if we add more types in the future.

`FLUME_TYPE_[THE_TYPE] = 2`

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,246 @@
+    @property
+    def unit_of_measurement(self):
+        """Return the unit the value is expressed in."""
+        return self._unit_of_measurement
```
As it's not going to change, you can return the value directly instead of storing it into a variable.

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,246 @@
+            return json.loads(response.text)["data"]
+        else:
+            raise Exception(
+                "getDevices Response Code not Successful. Returned {}".format(
```
Exception message should be updated for production.

Something like `Impossible to retreive devices. Response code returned : {}`

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,246 @@
+        self._client_secret = client_secret
+        self._device_id = device_id
+        self._scan_interval = scan_interval
+        self._time_zone = time_zone
```
Since you cleaned `getToken` + `getUserId`, some variables are useless here

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,246 @@
+            self.value = json.loads(response.text)["data"][0]["update"][0]["value"]
+        else:
+            raise Exception(
+                "getDevices Response Code not Successful. Returned {}".format(
```
Exception message should be updated for production.

Something like `Can't update flume data for user id {}. Response code returned : {}`

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,246 @@
+            return json.loads(response.text)["data"]
+        else:
+            raise Exception(
+                "getToken Response Code not Successful. Returned {}".format(
```
Exception message should be updated for production.

Something like `Can't get token for user "{}". Response code returned : {}` (self._username, response.status_code)

ChrisMandich committed:
Implemented suggested changes from @Quentame

This includes: components name, using f-strings instead of concat, snake_case for variables, constants for the addition of future device types, clearer errors, and removed variables no longer in use.

ChrisMandich commented on homeassistant/components/flume/manifest.json:
```diff
@@ -0,0 +1,9 @@
+{
+    "domain": "flume",
+    "name": "Flume, Chris Mandich",
```
This has been completed

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -84,15 +84,10 @@ def state(self):
-    @property
-    def unit_of_measurement(self):
-        """Return the unit the value is expressed in."""
-        return self._unit_of_measurement
```
I was thinking of `return "GALLONS"` instead of a variable (and I think it should be "gal" instead).

ChrisMandich replied:
Restored. I really appreciate all the feedback. Thank you!

ChrisMandich committed:
Update sensor.py

Restored unit_of_measurement. Updated return to "gal".

ChrisMandich committed:
Address pylint errors

ChrisMandich committed:
Update sensor.py

Include protected attributes in setup_platform.

ChrisMandich committed:
Address Pylint errors

homeassistant/components/flume/sensor.py:63:11: W0703: Catching too general exception Exception (broad-except)
homeassistant/components/flume/sensor.py:133:8: R1720: Unnecessary "else" after "raise" (no-else-raise)
homeassistant/components/flume/sensor.py:162:8: R1705: Unnecessary "else" after "return" (no-else-return)
homeassistant/components/flume/sensor.py:236:8: R1720: Unnecessary "else" after "raise" (no-else-raise)

ChrisMandich committed:
Update sensor.py

I'm okay with the broad exception clause.
homeassistant/components/flume/sensor.py:65:11: W0703: Catching too general exception Exception (broad-except)

ChrisMandich committed:
Update sensor.py

Add more specific exceptions for Try/Except.

ChrisMandich commented:
@Quentame I'm fairly new to contributing to Github projects. I believe I corrected all they Pylint errors. Can you guide me on what 'codecov/patch' is and what is required to get complete this successfully?

frenck commented:
All communication to external devices or services must be wrapped in an external Python library hosted on pypi. See the: <https://developers.home-assistant.io/docs/en/development_checklist.html>

Furthermore, the config flow requires tests, please add those to make `codecov` pass.

Lastly, please open up a PR in our documentation repository for this new integration. Please be sure to link the PR in the openings post of this PR.

Thanks! 👍

Quentame commented:
Hi again @ChrisMandich !

Me too, I'm quite new in contributions and also in Python dev actually. But most of the comments I've added to your PR are ones I already got while adding config flow to Linky and iCloud components.

I've haven't built a component from the ground up but your PR pushed me to review, hoping someone else will join the game 😉 (I'm not in HA team but new integration with passed pipeline attracts).

I think you should add your untested files to the `.coveragerc` file (be careful it is checked on the checklist but not made), if it's not that, I don't know 😬 

I know that config flow needs tests but this PR doesn't have it (you should consider adding it perhaps). 

@frenck : Where are the test/cov requirements written in the documentation ?

In the [component checklit](https://developers.home-assistant.io/docs/en/creating_component_code_review.html) there is nothing about the coverage and tests are just strongly recommended.

ChrisMandich committed:
Update Flume Sensory.py

add requirements, exclude from tests, us pyflume pypi package.

ChrisMandich commented:
Awesome, I appreciate the feedback. I have made the changes suggested and included the link to my PR for documentation found here: https://github.com/home-assistant/home-assistant.io/pull/10624

Added the component to `.coveragerc` file but will review adding tests in a future release.

frenck commented:
@Quentame @ChrisMandich I'm sorry, not sure where my head was at, but there is no config flow in this integration, so my remark regarding tests was incorrect.

Quentame commented on homeassistant/components/flume/manifest.json:
```diff
@@ -0,0 +1,9 @@
+    "documentation": "https://www.home-assistant.io/integrations/flume/",
+    "dependencies": [],
+    "codeowners": ["@ChrisMandich"],
+    "requirements": ["pytz==2019.2","pyflume==0.1.4"]
```
Nothing huge but I noticed most of the manifest.json are getting this order : 
domain
name
documentation
requirements
dependencies
codeowners

Can you put it the same way ?
And open requirements in multiple lines.

ChrisMandich replied:
Updated.

Quentame commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,94 @@
+    time_zone = str(hass.config.time_zone)
+    name = config[CONF_NAME]
+
+    flume_devices = FlumeAuth(_username, _password, _client_id, _client_secret)
```
Something weird here : you get flume devices with a FlumeAuth class.

It's good but if you can improve it will be nicer 😉

ChrisMandich replied:
Updated as well :D

Quentame reviewed (approved):
Wow, pretty clean sensor to me now !

Final review with @balloob or @MartinHjelmare ?

ChrisMandich committed:
Update sensor.py to support latest pyflume package

ChrisMandich committed:
Update manifest

provide better flow of manifest and easier readibility.

ChrisMandich commented:
@frenck In terms of Doc's is there anything additional I need to accomplish to update the tag?

balloob commented on homeassistant/components/flume/manifest.json:
```diff
@@ -0,0 +1,12 @@
+    "name": "Flume",
+    "documentation": "https://www.home-assistant.io/integrations/flume/",
+    "requirements": [
+        "pytz==2019.2",
```
No need to add pytz to requirements. It's already a core requirement.

ChrisMandich replied:
Done.

balloob commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,94 @@
+                    time_zone,
+                    SCAN_INTERVAL,
+                )
+                add_entities([FlumeSensor(flume, f"{name} {device['id']}")], True)
```
Please collect all entities in a list, then call `add_entities`

ChrisMandich replied:
Compelted

balloob commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,94 @@
+                add_entities([FlumeSensor(flume, f"{name} {device['id']}")], True)
+    except KeyError:
+        _LOGGER.error("No Flume Devices Returned of Type: %s", FLUME_TYPE_SENSOR)
+        return False
```
Platforms don't return booleans.

ChrisMandich replied:
Removed.

balloob commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,94 @@
+    try:
+        for device in flume_devices.device_list:
+            if device["type"] == FLUME_TYPE_SENSOR:
+                flume = FlumeData(
```
By creating a new object for each entity, you will also hit the server with a new API call on each update for each entity. So 10 entities is 10 API calls. This object should be shared and only 1 update should be made .

ChrisMandich replied:
I believe I implemented throttle which will help address this. The user_id and device_id is within the URI so I am unable to do one call for multiple devices.

ChrisMandich replied:
I'm not certain if this is resolved so I will leave it open until I receive confirmation from you.

balloob reviewed (changes requested):
We should rework how APIs are called. Look at the codebase and search for `@Throttle`.

ChrisMandich committed:
Update manifest.json

Reccomended by @balloob as it is already a core requirement

ChrisMandich committed:
Update sensor.py

Add proposed changes from @balloob

ChrisMandich committed:
Update requirements_all.txt

pytz is a core dependency, removing flume's requirement for it.

MartinHjelmare commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,98 @@
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Flume sensor."""
+    _username = config[CONF_USERNAME]
```
Please don't prefix local variables with underscore. It's not necessary.

ChrisMandich replied:
resolved

MartinHjelmare commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,98 @@
+                )
+                flume_entity_list.append(FlumeSensor(flume, f"{name} {device['id']}"))
+    except KeyError:
+        _LOGGER.error("No Flume Devices Returned of Type: %s", FLUME_TYPE_SENSOR)
```
Please just capitalize the first word in the logging message, not every word.

ChrisMandich replied:
resolved

MartinHjelmare commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,98 @@
+                flume_entity_list.append(FlumeSensor(flume, f"{name} {device['id']}"))
+    except KeyError:
+        _LOGGER.error("No Flume Devices Returned of Type: %s", FLUME_TYPE_SENSOR)
+    except AttributeError as attr_error:
```
When does the attribute error happen?

ChrisMandich replied:
After review, I don't believe this will occur. I have removed it as well. It originally was added as I was troubleshooting some of the automated code check's and my exception was too broad causing pylint to fail.

MartinHjelmare commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,98 @@
+        """Return the unit the value is expressed in."""
+        return "gal"
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES, MIN_TIME_BETWEEN_FORCED_UPDATES)
```
There are no forced updates, right?

ChrisMandich replied:
This is correct, I removed it as I believe it isn't necessary.

ChrisMandich committed:
Update sensor.py

Added @MartinHjelmare recommended changes.

MartinHjelmare commented on homeassistant/components/flume/sensor.py:
```diff
@@ -11,7 +11,7 @@
 from homeassistant.util import Throttle
 
-_LOGGER = logging.getLogger(__name__)
+Logger = logging.getLogger(__name__)
```
Please change this back to previous version.

ChrisMandich replied:
Resolved

MartinHjelmare commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,95 @@
+                    SCAN_INTERVAL,
+                )
+                flume_entity_list.append(FlumeSensor(flume, f"{name} {device['id']}"))
+    except KeyError:
```
Is there a chance that `"type"` key is missing in the device dict or just that the value of that key is not `FLUME_TYPE_SENSOR`?

ChrisMandich replied:
I have not seen an instance of the type key existing, but I do have an error to catch it. There is a chance that a `FLUME_TYPE_SENSOR` might not exist in the user's device list.

ChrisMandich replied:
Is there an action item you would like me to address for either of those occurring? @MartinHjelmare

MartinHjelmare replied:
I'm wondering why we are catching a `KeyError`. What key might be missing?

ChrisMandich replied:
Either `id` or `type`. Although if the flume api returns data this shouldn't be the case. Originally I had a more broad exception and narrowed it down to only key errors occurring in order to meet the automated tests.

ChrisMandich committed:
Update sensor.py

Resolving PyLint error

michaeldavie commented:
Thank for bringing this forward. I've been casually working on a similar [project](https://github.com/michaeldavie/flume-sensor), and I have a couple suggestions:

- My approach uses single data object to support multiple sensors, and uses a single API call with multiple queries to refresh them at the same time. It's similar to what many of the weather components do, and I think it makes sense for Flume as well.
- I've worked out a few queries/sensors to mirror the data provided in the native Flume app.
- I haven't completely worked out the OAuth2 support, and it looks like you might not have either. The tokens issued by Flume expire after a week, so we need to handle either refreshing a token or requesting a new one when the API returns an error indicating that the token is expired.

There's also a [thread](https://community.home-assistant.io/t/flume-water-meter/124485) that a few Flume users are following that might be useful.

Let me know if you have any questions about my code, or if there's anything I can do to help.

ChrisMandich committed:
Update sensor.py

Remove `KeyError`. Add length check for flume entity list before adding.

MartinHjelmare commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,93 @@
+            )
+            flume_entity_list.append(FlumeSensor(flume, f"{name} {device['id']}"))
+
+    if len(flume_entity_list) > 0:
```
```py
if flume_entity_list:
```

ChrisMandich committed:
Update sensor.py

MartinHjelmare commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,93 @@
+        """Return the unit the value is expressed in."""
+        return "gal"
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
```
This throttle will not throttle per home assistant but per entity. So each entity will be able to update with this time interval.

Usually a throttle is applied to a single instance that fetches the data for all entities, so that the throttle limits the api requests for the whole integration.

Is there any throttling in the library?

ChrisMandich replied:
I updated the library to use `ratelimit`. The API request URI is unique per device so I'm not sure how I would be able to have a single instance fetch all data when multiple devices are involved.

MartinHjelmare replied:
If the library has a throttle, we can remove the throttle here, right?

MartinHjelmare commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,93 @@
+import voluptuous as vol
+
+from pyflume import FlumeDeviceList, FlumeData
+from homeassistant.components.sensor import PLATFORM_SCHEMA
```
Please add a blank line between 3rd party and homeassistant import groups.

MartinHjelmare commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,93 @@
+import logging
+
+import voluptuous as vol
+
```
Please remove this blank line so that 3rd party imports are grouped together.

MartinHjelmare replied:
I recommend using the `isort` tool to sort and group imports automatically.

ChrisMandich replied:
Thank you for the recommendation!

ChrisMandich committed:
Update pyflume version

ChrisMandich committed:
Update imports with isort

MartinHjelmare commented on homeassistant/components/flume/sensor.py:
```diff
@@ -0,0 +1,92 @@
+"""Sensor for displaying the number of result from Flume."""
+from datetime import timedelta
+import logging
+from pyflume import FlumeData, FlumeDeviceList
```
There should be a blank line between standard library and 3rd party imports.

ChrisMandich committed:
Add line break between standard library and thirdparty imports.

ChrisMandich committed:
Remove throttle from sensor.py

MartinHjelmare reviewed (approved):
Looks good!

Quentame commented:
So it can be merged ?
